### PR TITLE
ci: protobuf linting (1/2): add buf config

### PR DESCRIPTION
**Pull request by nrainer-materialize**
Prerequisite for pull request https://github.com/MaterializeInc/materialize/pull/19157#discussion_r1188821306 of https://github.com/MaterializeInc/materialize/issues/17882.

### Motivation
* This is necessary for https://github.com/MaterializeInc/materialize/pull/19157 and needs to be merged before that.

### CODE REVIEW

**Review by def- — COMMENTED** [review]
Are we already using the protobuf linting tool or is the plan to add that later? Is this already enforced?

**Comment by nrainer-materialize** [comment]
> Are we already using the protobuf linting tool or is the plan to add that later? Is this already enforced?

It will be introduced with the follow-up PR, which is https://github.com/MaterializeInc/materialize/pull/19157.

**Review by teskje — COMMENTED** [review]
LGTM, with the caveat that I don't know whether this is the right place to put this config file. We seem to usually place them one level higher in the crate root, at the moment `src/` only contains crate directories. @benesch might have a take here.

**Comment by teskje** [comment]
If it's hard to move the config file anywhere else then I'm fine with leaving it there. At least I don't see a reason why it must not reside inside `src/`. I was just assuming that it would be automatically picked up in parent directories as well.

**Comment by benesch** [comment]
> If it's hard to move the config file anywhere else then I'm fine with leaving it there. At least I don't see a reason why it must not reside inside `src/`. I was just assuming that it would be automatically picked up in parent directories as well.

It looks like buf.yml must go at the root of the protobuf hierarchy. We use `src` as the root of our proto hierarchy today. It's a bit hard to see this, but you can back it out from the fact that every build.rs file in every crate specifies ".." as the include path

https://github.com/MaterializeInc/materialize/blob/b501b008d95c019cbe85401e1b71eb52209cb331/src/repr/build.rs#L104

which resolves to `src`.

Could totally change this to be the root of the repo ("../.."), but doing so would be a big yak shave, as you'd need to update all the build.rs files _and_ update every .proto's imports to say "src/" at the start of the path.

**Comment by benesch** [comment]
tl;dr all good with me to have this inside `src` for now, but happy for @teskje or someone else who cares more about proto to assert that we should do the work to make `/` our proto root, rather than `/src`.

**Review by def- — APPROVED** [review]
Ok for me, but I think this could also just be part of https://github.com/MaterializeInc/materialize/pull/19157, then we can immediately see if the config results in any problems or not.

**Comment by teskje** [comment]
> tl;dr all good with me to have this inside `src` for now, but happy for @teskje or someone else who cares more about proto to assert that we should do the work to make `/` our proto root, rather than `/src`.

I don't mind either way. CCing @aalexandrov as the person who is most likely to care about proto.

**Review by teskje — APPROVED** [review]

**Comment by aalexandrov** [comment]
> CCing @aalexandrov as the person who is most likely to care about proto.

I don't have strong objections against that, we should do whatever is in line with the current repository layout, as long as we find somebody to to the `sed` / `regex` work mentioned by @benesch:

> Could totally change this to be the root of the repo ("../.."), but doing so would be a big yak shave, as you'd need to update all the build.rs files and update every .proto's imports to say "src/" at the start of the path.

**Review by philip-stoev — APPROVED** [review]

File: src/buf.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+version: v1
+breaking:
+  use:
+    - FILE
+lint:
+  use:
+    - DEFAULT
+build:
+  excludes:
```
Comment (philip-stoev) [inline src/buf.yaml:18]: A comment would be nice as to why those files are excluded -- I believe the original ticket does list the reason , so just copy it over as a comment here if you do not mind.

Comment (nrainer-materialize) [reply src/buf.yaml:18]: ✅ Done.
